src/java.base/share/classes/java/security/PEMDecoder.java

```diff
@@ -0,0 +1,486 @@
+import java.util.Base64;
+import java.util.Objects;
+
+/**
+ * {@code PEMDecoder} implements a decoder for Privacy-Enhanced Mail (PEM) data.
+ * PEM is a textual encoding used to store and transfer security
+ * objects, such as asymmetric keys, certificates, and certificate revocation
+ * lists (CRLs).  It is defined in RFC 1421 and RFC 7468.  PEM consists of a
+ * Base64-formatted binary encoding enclosed by a type-identifying header
+ * and footer.
+ *
+ * <p> The {@linkplain #decode(String)} and {@linkplain #decode(InputStream)}
+ * methods return an instance of a class that matches the data
+ * type and implements {@link DEREncodable}. The following types are
+ * decoded into Java API cryptographic object that implement
+ * {@link DEREncodable}:
+ * <pre>
+ *     PRIVATE KEY, RSA PRIVATE KEY, PUBLIC KEY, CERTIFICATE,
+ *     X509 CERTIFICATE, X509 CRL, and ENCRYPTED PRIVATE KEY.
+ * </pre>
+ *
+ * <p> If the PEM does not have a JCE object representation, it returns a
+ * {@link PEMRecord}.
+ *
+ * <p> The {@linkplain #decode(String, Class)} and
+ * {@linkplain #decode(InputStream, Class)} methods take a Class parameter
+ * which determines the type of {@code DEREncodable} that is returned. These
+ * methods are useful when extracting or changing the return class.
+ * The Class parameter can specify the returned
+ * key object from a PEM containing a public and private key.  If only
+ * the private key is required, {@code PrivateKey.class} can be used.
+ * If the Class parameter is set to {@code X509EncodedKeySpec.class}, the
+ * public key will be returned in that format.  Any type of PEM data can be
+ * decoded into a {@code PEMRecord} by specifying {@code PEMRecord.class}.
+ * If the Class parameter doesn't match the PEM content, an
+ * {@code IllegalArgumentException} will be thrown.
+ *
+ * <p> A new {@code PEMDecoder} instance is created when configured
+ * with {@linkplain #withFactory(Provider)} and/or
+ * {@linkplain #withDecryption(char[])}. {@linkplain #withFactory(Provider)}
+ * configures the decoder to use only {@linkplain KeyFactory} and
+ * {@linkplain CertificateFactory} instances from the given {@code Provider}.
+ * {@link#withDecryption(char[])} configures the decoder to decrypt all
+ * encrypted private key PEM data using the given password.
+ * Configuring an instance for decryption does not prevent decoding with
+ * unencrypted PEM. Any encrypted PEM that does not use the configured password
+ * will throw a {@link RuntimeException}. When encrypted PEM is used with a
+ * decoder not configured for decryption, an {@link EncryptedPrivateKeyInfo}
+ * object is returned.  {@code EncryptedPrivateKeyInfo} methods must be used to
+ * retrieve the {@link PrivateKey}.
+ *
+ * <p>This class is immutable and thread-safe.
+ *
+ * <p> Here is an example of decoding a {@code PrivateKey} object:
+ * {@snippet lang = java:
+ *     PEMDecoder pd = PEMDecoder.of();
+ *     PrivateKey priKey = pd.decode(priKeyPEM, PrivateKey.class);
+ * }
```

Comment:
Add examples on `withFactory` and `withDecryption`.

Reply:
ok